### PR TITLE
fix: format throw data truncated error

### LINKED ISSUE #4891: `format` throw data truncated error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> create table format_test(id int, value decimal(7,4));
Query OK, 0 rows affected (0.09 sec)

mysql> insert into format_test values(1,999.9999);
Query OK, 1 row affected (0.02 sec)

mysql> select format(value,0) from format_test;
+-----------------+
| format(value,0) |
+-----------------+
| 1,000           |
+-----------------+
1 row in set (0.00 sec)

mysql> alter table format_test set tiflash replica 1;
Query OK, 0 rows affected (0.09 sec)
mysql> set tidb_enforce_mpp=1;
Query OK, 0 rows affected (0.00 sec)

mysql> explain select format(value,0) from format_test;
+---------------------------+----------+--------------+-------------------+---------------------------------------------+
| id                        | estRows  | task         | access object     | operator info                               |
+---------------------------+----------+--------------+-------------------+---------------------------------------------+
| TableReader_11            | 10000.00 | root         |                   | data:ExchangeSender_10                      |
| └─ExchangeSender_10       | 10000.00 | mpp[tiflash] |                   | ExchangeType: PassThrough                   |
|   └─Projection_4          | 10000.00 | mpp[tiflash] |                   | format(test.format_test.value, 0)->Column#4 |
|     └─TableFullScan_9     | 10000.00 | mpp[tiflash] | table:format_test | keep order:false, stats:pseudo              |
+---------------------------+----------+--------------+-------------------+---------------------------------------------+
4 rows in set (0.00 sec)

mysql>  select format(value,0) from format_test;
ERROR 1105 (HY000): other error for mpp stream: DB::TiFlashException: Data truncated

```
### 2. What did you expect to see? (Required)
```
mysql> select format(value,0) from format_test;
+-----------------+
| format(value,0) |
+-----------------+
| 1,000           |
+-----------------+

```
### 3. What did you see instead (Required)
```
mysql>  select format(value,0) from format_test;
ERROR 1105 (HY000): other error for mpp stream: DB::TiFlashException: Data truncated

```
### 4. What is your TiFlash version? (Required)
master @ b8e58ebcb741e3a921fc93352de72c5a20a5ba6f

### PR DESCRIPTION
**Pull request by xzhangxian1008**
### What problem does this PR solve?

Issue Number: close #4891

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Manual Test:
```
mysql> select * from test;
+------+--------------------------------------------------------------------+
| id   | value                                                              |
+------+--------------------------------------------------------------------+
|    1 | 9999999999999999999999999999999999999999999999999999999999999.9999 |
+------+--------------------------------------------------------------------+

mysql> select format(value, -3) from test;
+------------------------------------------------------------------------------------+
| format(value, -3)                                                                  |
+------------------------------------------------------------------------------------+
| 10,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000 |
+------------------------------------------------------------------------------------+

mysql> select format(value, 0) from test;
+------------------------------------------------------------------------------------+
| format(value, 0)                                                                   |
+------------------------------------------------------------------------------------+
| 10,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000 |
+------------------------------------------------------------------------------------+

mysql> select format(value, 3) from test;
+----------------------------------------------------------------------------------------+
| format(value, 3)                                                                       |
+----------------------------------------------------------------------------------------+
| 10,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000,000.000 |
+----------------------------------------------------------------------------------------+

mysql> select format(value, 10) from test;
+----------------------------------------------------------------------------------------------+
| format(value, 10)                                                                            |
+----------------------------------------------------------------------------------------------+
| 9,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999,999.9999000000 |
+----------------------------------------------------------------------------------------------+

mysql> explain select format(value, 10) from test;
+---------------------------+---------+--------------+---------------+---------------------------------------+
| id                        | estRows | task         | access object | operator info                         |
+---------------------------+---------+--------------+---------------+---------------------------------------+
| TableReader_11            | 1.00    | root         |               | data:ExchangeSender_10                |
| └─ExchangeSender_10       | 1.00    | mpp[tiflash] |               | ExchangeType: PassThrough             |
|   └─Projection_4          | 1.00    | mpp[tiflash] |               | format(test.test.value, 10)->Column#4 |
|     └─TableFullScan_9     | 1.00    | mpp[tiflash] | table:test    | keep order:false, stats:pseudo        |
+---------------------------+---------+--------------+---------------+---------------------------------------+
```

```
mysql> select * from format_test;
+------+----------+
| id   | value    |
+------+----------+
|    1 | 999.9999 |
+------+----------+

mysql> select format(value, -2) from format_test;
+-------------------+
| format(value, -2) |
+-------------------+
| 1,000             |
+-------------------+

mysql> select format(value, 0) from format_test;
+------------------+
| format(value, 0) |
+------------------+
| 1,000            |
+------------------+

mysql> select format(value, 2) from format_test;
+------------------+
| format(value, 2) |
+------------------+
| 1,000.00         |
+------------------+

mysql> select format(value, 10) from format_test;
+-------------------+
| format(value, 10) |
+-------------------+
| 999.9999000000    |
+-------------------+

mysql> explain select format(value, 10) from format_test;
+---------------------------+---------+--------------+-------------------+----------------------------------------------+
| id                        | estRows | task         | access object     | operator info                                |
+---------------------------+---------+--------------+-------------------+----------------------------------------------+
| TableReader_11            | 1.00    | root         |                   | data:ExchangeSender_10                       |
| └─ExchangeSender_10       | 1.00    | mpp[tiflash] |                   | ExchangeType: PassThrough                    |
|   └─Projection_4          | 1.00    | mpp[tiflash] |                   | format(test.format_test.value, 10)->Column#4 |
|     └─TableFullScan_9     | 1.00    | mpp[tiflash] | table:format_test | keep order:false, stats:pseudo               |
+---------------------------+---------+--------------+-------------------+----------------------------------------------+


```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Fix the issue that format throw data truncated error
```

### CODE REVIEW

**Comment by SeaRise** [comment]
enable ut at https://github.com/pingcap/tiflash/blob/045d24fe854193ff7fe9e56f0ed1ff0e4aa1cddc/dbms/src/Functions/tests/gtest_strings_format.cpp in this pr?

**Comment by SeaRise** [comment]
and add release note: Fix the issue that format throw data truncated error

File: dbms/src/Functions/FunctionsRound.h [hunk 1]
```diff
@@ -1186,7 +1201,7 @@ struct TiDBDecimalRound
             scaled_value *= PowForOutput::result[-difference];
 
         // check overflow and construct result.
-        if (scaled_value > DecimalMaxValue::get(info.output_prec))
+        if (scaled_value > DecimalMaxValue::get(info.output_prec + carry))
```
Comment (fuzhe1989) [inline dbms/src/Functions/FunctionsRound.h:1204]: IMO it's safe to always take carry into accout: `if (scaled_value > DecimalMaxValue::get(info.output_prec + 1))`.

Comment (windtalker) [reply dbms/src/Functions/FunctionsRound.h:1204]: I aggree, just make sure `info.output_prec + 1 <= max_decimal_prec`

**Comment by xzhangxian1008** [comment]
> enable ut at https://github.com/pingcap/tiflash/blob/045d24fe854193ff7fe9e56f0ed1ff0e4aa1cddc/dbms/src/Functions/tests/gtest_strings_format.cpp in this pr?

Done

**Comment by xzhangxian1008** [comment]
> Please add some tests, and enable the existing test [here](https://github.com/pingcap/tiflash/blob/6da631c99c918bfffcf183128306a5e6bd35c7f7/dbms/src/Functions/tests/gtest_strings_format.cpp#L65)

Done

File: dbms/src/Functions/FunctionsString.cpp [hunk 2]
```diff
@@ -4573,7 +4576,14 @@ class FormatImpl : public IFunction
             using NumberFieldType = typename NumberType::FieldType;
             using NumberColVec = std::conditional_t<IsDecimal<NumberFieldType>, ColumnDecimal<NumberFieldType>, ColumnVector<NumberFieldType>>;
             const auto * number_raw = block.getByPosition(arguments[0]).column.get();
-            TiDBDecimalRoundInfo info{number_type, number_type};
+
+            auto prec = getDecimalPrecision(number_type, 0);
+            auto scale = getDecimalScale(number_type, 0);
+
+            /// output_scale is the second parameter of the 'format' function
+            auto output_scale = precision_base_col_size > 0 ? (*precision_base_col)[0].get<Int64>() : scale;
```
Comment (windtalker) [inline dbms/src/Functions/FunctionsString.cpp:4584]: What happens if `precision_base_col` is not a constant column, in this case the output_scale is differ from row to row.

Comment (xzhangxian1008) [reply dbms/src/Functions/FunctionsString.cpp:4584]: > What happens if `precision_base_col` is not a constant column, in this case the output_scale is differ from row to row.

When `precision_base_col` is not a constant column, we will always reset the `info.output_scale` with `max_num_decimals`  which can be shown in FunctionsString.cpp:4625.

**Comment by SeaRise** [comment]
> > enable ut at https://github.com/pingcap/tiflash/blob/045d24fe854193ff7fe9e56f0ed1ff0e4aa1cddc/dbms/src/Functions/tests/gtest_strings_format.cpp in this pr?
> 
> Done

emmm...

**Comment by xzhangxian1008** [comment]
> > > enable ut at https://github.com/pingcap/tiflash/blob/045d24fe854193ff7fe9e56f0ed1ff0e4aa1cddc/dbms/src/Functions/tests/gtest_strings_format.cpp in this pr?
> > 
> > 
> > Done
> 
> emmm... I mean the following two cases
> 
> https://github.com/pingcap/tiflash/blob/e9315215b4ff6064323a2fb53f6db4374b9b6bd8/dbms/src/Functions/tests/gtest_strings_format.cpp#L114-L135
> 
> 
> https://github.com/pingcap/tiflash/blob/6da631c99c918bfffcf183128306a5e6bd35c7f7/dbms/src/Functions/tests/gtest_strings_format.cpp#L111-L131

okkk, I have enabled them

**Review by windtalker — APPROVED** [review]
LGTM

**Review by SeaRise — APPROVED** [review]
/merge

**Pull request merged**